plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java

```diff
@@ -93,19 +96,17 @@ public AmazonS3Reference client(String clientName) {
         }
     }
 
-    private AmazonS3 buildClient(S3ClientSettings clientSettings) {
-        final AWSCredentialsProvider credentials = buildCredentials(logger, clientSettings);
-        final ClientConfiguration configuration = buildConfiguration(clientSettings);
-        final AmazonS3 client = buildClient(credentials, configuration);
-        if (Strings.hasText(clientSettings.endpoint)) {
-            client.setEndpoint(clientSettings.endpoint);
-        }
-        return client;
-    }
-
     // proxy for testing
-    AmazonS3 buildClient(AWSCredentialsProvider credentials, ClientConfiguration configuration) {
-        return new AmazonS3Client(credentials, configuration);
+    AmazonS3 buildClient(final S3ClientSettings clientSettings) {
+        final AmazonS3ClientBuilder builder = AmazonS3ClientBuilder.standard();
+        builder.withCredentials(buildCredentials(logger, clientSettings));
+        builder.withClientConfiguration(buildConfiguration(clientSettings));
+
+        final String endpoint = Strings.hasLength(clientSettings.endpoint) ? clientSettings.endpoint : Constants.S3_HOSTNAME;
+        logger.debug(() -> new ParameterizedMessage("using endpoint [{}]", endpoint));
+        builder.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endpoint, null));
```

Comment:
It took me some time to see what was happening here.

If the endpoint configuration isn't set on the _builder_ then the default behaviour is to try and work out what region we are in and use an appropriate endpoint - see `AwsClientBuilder#setRegion`. In contrast, directly-constructed clients use `s3.amazonaws.com` unless otherwise instructed. We currently use a directly-constructed client, and need to keep the existing behaviour to avoid a breaking change, so to move to using the builder we must set it explicitly to keep the existing behaviour.

AIUI we do this because directly constructing the client is deprecated (was already deprecated in 1.1.223 too) so this change removes that usage of a deprecated API.

A short comment to this effect would be useful, I think, to save future readers.

Reply:
I presume, by the way, that passing `null` for the region here is ok. I haven't tried it.

Reply:
Thanks for the suggested comment, I'll steal it from you.

> I presume, by the way, that passing null for the region here is ok. I haven't tried it.

Yes. Similarly to the deprecated `setEndpoint()` method that leaves the region to null until it is later resolved from the endpoint to the default us-east-1 region.